#**Infopillar Solutions Pvt.Ltd**
### **Task 1- Fake News Detection Project**
### *Fake news spreads like a wildfire and this is a big issue in this era.So taking a small step to avoid such news.*

 **Author : Mr. SIJU SAJI**

# **Import Libraries**

In [125]:
import numpy as np
import pandas as pd
import itertools
from sklearn.model_selection import train_test_split
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.linear_model import PassiveAggressiveClassifier
from sklearn.metrics import accuracy_score, confusion_matrix

# **Loading Dataset**
Here the Dataset is extracted from the drive

In [126]:
from google.colab import drive
drive.mount('/content/drive')
news=pd.read_csv("/news.csv")
news

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


Unnamed: 0  ... label
0           8476  ...  FAKE
1          10294  ...  FAKE
2           3608  ...  REAL
3          10142  ...  FAKE
4            875  ...  REAL
...          ...  ...   ...
6330        4490  ...  REAL
6331        8062  ...  FAKE
6332        8622  ...  FAKE
6333        4021  ...  REAL
6334        4330  ...  REAL

[6335 rows x 4 columns]

# **Summarise The dataset**

In [127]:
print("Number of rows and  columns:",news.shape)

Number of rows and  columns: (6335, 4)


In [128]:
print("The first five rows are:",news.head())

The first five rows are:    Unnamed: 0  ... label
0        8476  ...  FAKE
1       10294  ...  FAKE
2        3608  ...  REAL
3       10142  ...  FAKE
4         875  ...  REAL

[5 rows x 4 columns]


# **Presenting the Labels from Dataset**

In [129]:
labels=news.label
labels

0       FAKE
1       FAKE
2       REAL
3       FAKE
4       REAL
        ... 
6330    REAL
6331    FAKE
6332    FAKE
6333    REAL
6334    REAL
Name: label, Length: 6335, dtype: object

In [130]:
news.tail()

Unnamed: 0  ... label
6330        4490  ...  REAL
6331        8062  ...  FAKE
6332        8622  ...  FAKE
6333        4021  ...  REAL
6334        4330  ...  REAL

[5 rows x 4 columns]

# **Train-Test Split Evaluation**

In [131]:
x_train,x_test,y_train,y_test=train_test_split(news['text'], labels, test_size=0.63, random_state=8)

In [132]:
tfidf_vectorizer=TfidfVectorizer(stop_words='english',max_df=0.8)

# **Fit and transform the vectorizer on the train set**
# **And transform the vectorizer on the test set**

In [133]:
tfidf_train=tfidf_vectorizer.fit_transform(x_train) 
tfidf_test=tfidf_vectorizer.transform(x_test)

# **Initializing a Passive_Aggressive_Classifier**

In [134]:
pac=PassiveAggressiveClassifier(max_iter=50)
pac.fit(tfidf_train,y_train)

PassiveAggressiveClassifier(C=1.0, average=False, class_weight=None,
                            early_stopping=False, fit_intercept=True,
                            loss='hinge', max_iter=50, n_iter_no_change=5,
                            n_jobs=None, random_state=None, shuffle=True,
                            tol=0.001, validation_fraction=0.1, verbose=0,
                            warm_start=False)

# **Predict on the test set and calculate accuracy**

In [135]:
y_pred=pac.predict(tfidf_test)
score=accuracy_score(y_test,y_pred)
print(f'Accuracy: {round(score*100,2)}%')

Accuracy: 92.31%


# **Here we could see that the accuracy is** ***92.31%*** (The accuracy might vary sometimes)

In [136]:
confusion_matrix(y_test,y_pred, labels=['FAKE','REAL'])

array([[1844,  134],
       [ 173, 1841]])

## **At the end we could see that , We have**
# ***1840 True Positives/***
# ***1832 True Negatives/***
# ***138 False Positives/***
# ***182 False Negatives***